### ISSUE TITLE
Exclude images in subfolders

### ISSUE THREAD

**Issue by X-Ray-Jin** [crashinvaders/gdx-texture-packer-gui#56]
Hey there, first of all I really like this GUI Packer. Good job and thanks for your hard work.
It really suits our needs very well. There is just one thing about the exclude function of images. I kind of wonder what the purpose of it is if it is not a bug:
My use case is to exclude some pictures within an input folder or even a subfolder of it. I added an input folder and when I saw the exclude function I intuitively expected it to ... exclude the images listed from the final atlas. However this does not seem to work like this. I wonder if this is a bug because if I dont want images to be packed when adding them one by one...well I just don't add them. Sure it makes de-/activating some images temporarily more easy but in 3 years of game dev I never would have needed that. If they are not needed at all we just remove em all together from the folder. Even if it is just temporarily the effort is minimal. So I think a really huge feature would be to exclude specific images or complete subfolders when using input folders (maybe even regular expressions for name excludes)

To put it more into perspective:
We are using spine and to make workflow more easy I would like to have a "raw" assets folder that the animators can use as root folder in spine and where they can find anything they need for the animations without messing with path hierarchy. However there are some utility images (alignment, canvas, etc.) or icons (Level editor) that are not used ingame but have to be in the hierachy so they can be found within the animation software without messing up the paths inside exported files. I.e when loading animations and levels ingame the paths of spine export file and the atlas have to match. We collect all single images from spine animations instead of using an atlas for each animation. This is to minimize texture swaps which is a great perfomance issue.

It would be nice to hear your thoughts on this.
Greetz X-Ray-Jin

**Comment by metaphore** [comment]
Hey mate, thanks for the input!

First of all, you're right, the exclude feature is mostly here to work together with input folders. Just to let you keep some specific images out of the atlas, but still keep them in the folder. I should say that I work with Spine on a daily basis as well, and implemented _exclude files_ just for the case when I need to have some _utility_ images, just like you described, so it makes perfect sense. Probably having regexp stuff in here would be a bit too much for now, but I like the idea about _excluded folders_, just because we have _recursive_ option for input dirs, this could improve the experience.

But lets get back to an issue, you say you have input folder and excluded file in it, but the atlas keep having that excluded image, right? If that's the case, could you please assemble a simple `.tpproj` with all input hierarchy that demonstrates the malfunction and share with me?

**Comment by X-Ray-Jin** [comment]
Yea exactly that is what happens. Sure here is the project file.

[global_character.zip](https://github.com/crashinvaders/gdx-texture-packer-gui/files/2311665/global_character.zip)

Concerning Regex yea i know. Just wanted to mention it as a kind of epic feature ^^ But maybe a simple contains string thing would be quite fast and easy and also kind of powerful as a first step.

**Comment by tommyettinger** [comment]
Regex filters could be a little tricky for users because `.` has special meaning in regexes and is also part of almost every file name, like `Jim_Walk_Left.png`. There's also file glob syntax that is sometimes used, and I think multiple "flavors" of glob syntax exist... I think the solution to most input file setup is to have the folders laid out exactly how the application expects them to be in the atlas, and to write small scripts if you need to move or copy things into other locations (using bash, shell, Python, Perl, Java even... whatever you know). I don't use Spine, though, so I don't know how utility images work.

I'd paste my huge many-file project with no exclusions, but I don't think there's much need. It's 6205 files in 102 folders, and I believe I copied some files into or out of the hierarchy using the Windows tool `robocopy.exe`, since it is one of the fastest ways to move files on Windows.

**Comment by tommyettinger** [comment]
Opening that `.tpproj` file, it looked like the excluded directories have different root folders than the included directory, but on closer inspection the path is just being shortened. Maybe gdx-texture-packer-gui expects the exclusions to be relative to some included file/folder? There's no files to check against though, so I can't see what it would produce.

**Comment by X-Ray-Jin** [comment]
> Regex filters could be a little tricky for users because . has special meaning in regexes and is also part of almost every file name, like Jim_Walk_Left.png

Regex sure are not common use for some users but for advanced users it could be quite handy. But you are right *.png could be a little tricky. On the other hand: shouldnt it be possible to exclude file endings from check?

What exactly you mean by:
> There's no files to check against though, so I can't see what it would produce.
I included the output file and these 2 pink squares you see in it are not supposed to be in there. If that is what you mean. (I uploaded a new zip maybe you got the old one)

Dont how to add relative to the input folder with the given UI...Looking at the file in text edit all paths are relativ to the project file.
(might be useful to also show that in the UI speaking of shortened paths for convience, just an idea)
